File: Sources/ClaudeOSRuntime/LoopStore.swift
import Foundation
import GRDB

/// SQLite-backed store for loop tasks and their proof ledger (`loops.sqlite`). Separate file
/// from the index and the brain: this is the orchestration layer's own state on disk, so a
/// loop can be reviewed, re-run, or resumed across app restarts. Explicit column mapping (UUID
/// as text, dates as doubles) keeps the schema predictable and queryable.
///
/// `@unchecked Sendable` is safe: the only stored property is a `DatabaseQueue`, which
/// serializes all access through its own queue.
public final class LoopStore: @unchecked Sendable {
    private let dbQueue: DatabaseQueue

    public init(path: String) throws {
        var config = Configuration()
        config.busyMode = .timeout(5)
        config.prepareDatabase { db in
            try db.execute(sql: "PRAGMA journal_mode = WAL")
            try db.execute(sql: "PRAGMA foreign_keys = ON")
        }
        dbQueue = try DatabaseQueue(path: path, configuration: config)
        try migrator.migrate(dbQueue)
    }

    private var migrator: DatabaseMigrator {
        var m = DatabaseMigrator()
        m.registerMigration("v1") { db in
            try db.create(table: "loop_task") { t in
                t.primaryKey("id", .text)
                t.column("title", .text).notNull()
                t.column("prompt", .text).notNull()
                t.column("cwd", .text).notNull()
                t.column("doneWhen", .text).notNull()
                t.column("checkKind", .text).notNull()
                t.column("maxPasses", .integer).notNull()
                t.column("rememberOnPass", .integer).notNull()
                t.column("state", .text).notNull()
                t.column("lastAttempt", .integer).notNull()
                t.column("createdAt", .double).notNull()
                t.column("updatedAt", .double).notNull()
            }
            try db.create(table: "proof") { t in
                t.primaryKey("id", .text)
                t.column("taskId", .text).notNull()
                    .references("loop_task", onDelete: .cascade)
                t.column("attempt", .integer).notNull()
                t.column("makerOutput", .text).notNull()
                t.column("passed", .integer).notNull()
                t.column("checkerOutput", .text).notNull()
                t.column("date", .double).notNull()
            }
            try db.create(indexOn: "proof", columns: ["taskId"])
        }
        m.registerMigration("v2-makerSessionId") { db in
            try db.alter(table: "proof") { t in
                t.add(column: "makerSessionId", .text)
            }
        }
        return m
    }

    // MARK: - Tasks

    public func allTasks() throws -> [LoopTask] {
        try dbQueue.read { db in
            try Row.fetchAll(db, sql: "SELECT * FROM loop_task ORDER BY updatedAt DESC").map(Self.task)
        }
    }

    public func upsert(_ task: LoopTask) throws {
        try dbQueue.write { db in
            try db.execute(sql: """
                INSERT INTO loop_task
                  (id, title, prompt, cwd, doneWhen, checkKind, maxPasses, rememberOnPass, state, lastAttempt, createdAt, updatedAt)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?)
                ON CONFLICT(id) DO UPDATE SET
                  title=excluded.title, prompt=excluded.prompt, cwd=excluded.cwd, doneWhen=excluded.doneWhen,
                  checkKind=excluded.checkKind, maxPasses=excluded.maxPasses, rememberOnPass=excluded.rememberOnPass,
                  state=excluded.state, lastAttempt=excluded.lastAttempt, updatedAt=excluded.updatedAt
                """,
                arguments: [task.id.uuidString, task.title, task.prompt, task.cwd, task.doneWhen,
                            task.checkKind.rawValue, task.maxPasses, task.rememberOnPass ? 1 : 0,
                            task.state.rawValue, task.lastAttempt,
                            task.createdAt.timeIntervalSinceReferenceDate,
                            task.updatedAt.timeIntervalSinceReferenceDate])
        }
    }

    public func deleteTask(id: UUID) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM loop_task WHERE id = ?", arguments: [id.uuidString])
        }
    }

    // MARK: - Proof ledger

    public func insert(proof: ProofRecord) throws {
        try dbQueue.write { db in
            try db.execute(sql: """
                INSERT INTO proof (id, taskId, attempt, makerOutput, passed, checkerOutput, makerSessionId, date)
                VALUES (?,?,?,?,?,?,?,?)
                """,
                arguments: [proof.id.uuidString, proof.taskId.uuidString, proof.attempt,
                            proof.makerOutput, proof.passed ? 1 : 0, proof.checkerOutput,
                            proof.makerSessionId, proof.date.timeIntervalSinceReferenceDate])
        }
    }

    public func proofs(taskId: UUID) throws -> [ProofRecord] {
        try dbQueue.read { db in
            try Row.fetchAll(db, sql: "SELECT * FROM proof WHERE taskId = ? ORDER BY attempt ASC",
                             arguments: [taskId.uuidString]).map(Self.proof)
        }
    }

    public func clearProofs(taskId: UUID) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM proof WHERE taskId = ?", arguments: [taskId.uuidString])
        }
    }

    // MARK: - Row mapping

    private static func task(_ r: Row) -> LoopTask {
        LoopTask(
            id: UUID(uuidString: r["id"]) ?? UUID(),
            title: r["title"], prompt: r["prompt"], cwd: r["cwd"], doneWhen: r["doneWhen"],
            checkKind: CheckKind(rawValue: r["checkKind"]) ?? .agent,
            maxPasses: r["maxPasses"], rememberOnPass: (r["rememberOnPass"] as Int) != 0,
            state: LoopState(rawValue: r["state"]) ?? .idle,
            lastAttempt: r["lastAttempt"],
            createdAt: Date(timeIntervalSinceReferenceDate: r["createdAt"]),
            updatedAt: Date(timeIntervalSinceReferenceDate: r["updatedAt"]))
    }

    private static func proof(_ r: Row) -> ProofRecord {
        ProofRecord(
            id: UUID(uuidString: r["id"]) ?? UUID(),
            taskId: UUID(uuidString: r["taskId"]) ?? UUID(),
            attempt: r["attempt"], makerOutput: r["makerOutput"],
            passed: (r["passed"] as Int) != 0, checkerOutput: r["checkerOutput"],
            makerSessionId: r["makerSessionId"],
            date: Date(timeIntervalSinceReferenceDate: r["date"]))
    }
}

File: windows-winui/Core/LoopStore.cs
using System.IO;
using Microsoft.Data.Sqlite;

namespace SiftWinUI.Core;

public enum CheckKind { Agent, Shell }

public sealed class LoopTask
{
    public string Id = Guid.NewGuid().ToString("N");
    public string Title = "";
    public string Prompt = "";
    public string Cwd = "";
    public string DoneWhen = "";
    public CheckKind Check = CheckKind.Agent;
    public int MaxPasses = 3;
    public string State = "idle";       // idle | running | checking | passed | failed | stopped
    public int LastAttempt;
    public long UpdatedAt;
}

public sealed record Proof(string TaskId, int Attempt, bool Passed, string MakerOutput,
                           string CheckerOutput, long At);

/// Loop definitions and their evidence, in their own database beside the index so a
/// rebuilt index never takes the record of what was run with it.
public sealed class LoopStore : IDisposable
{
    private readonly SqliteConnection _db;

    public LoopStore(string path)
    {
        var dir = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);
        _db = new SqliteConnection($"Data Source={path}");
        _db.Open();
        using var cmd = _db.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS loop_task (
              id TEXT PRIMARY KEY, title TEXT NOT NULL, prompt TEXT NOT NULL, cwd TEXT NOT NULL,
              doneWhen TEXT NOT NULL, checkKind TEXT NOT NULL, maxPasses INTEGER NOT NULL,
              state TEXT NOT NULL, lastAttempt INTEGER NOT NULL, updatedAt INTEGER NOT NULL);
            CREATE TABLE IF NOT EXISTS proof (
              taskId TEXT NOT NULL, attempt INTEGER NOT NULL, passed INTEGER NOT NULL,
              makerOutput TEXT NOT NULL, checkerOutput TEXT NOT NULL, at INTEGER NOT NULL);
            CREATE INDEX IF NOT EXISTS proof_task ON proof(taskId);
            """;
        cmd.ExecuteNonQuery();
    }

    public void Dispose() => _db.Dispose();

    private SqliteCommand Cmd(string sql, params object?[] args)
    {
        var cmd = _db.CreateCommand();
        cmd.CommandText = sql;
        for (var i = 0; i < args.Length; i++) cmd.Parameters.AddWithValue($"@p{i}", args[i] ?? DBNull.Value);
        return cmd;
    }

    public List<LoopTask> All()
    {
        var list = new List<LoopTask>();
        using var cmd = Cmd("SELECT * FROM loop_task ORDER BY updatedAt DESC");
        using var r = cmd.ExecuteReader();
        while (r.Read())
        {
            list.Add(new LoopTask
            {
                Id = r.GetString(r.GetOrdinal("id")),
                Title = r.GetString(r.GetOrdinal("title")),
                Prompt = r.GetString(r.GetOrdinal("prompt")),
                Cwd = r.GetString(r.GetOrdinal("cwd")),
                DoneWhen = r.GetString(r.GetOrdinal("doneWhen")),
                Check = r.GetString(r.GetOrdinal("checkKind")) == "shell" ? CheckKind.Shell : CheckKind.Agent,
                MaxPasses = r.GetInt32(r.GetOrdinal("maxPasses")),
                State = r.GetString(r.GetOrdinal("state")),
                LastAttempt = r.GetInt32(r.GetOrdinal("lastAttempt")),
                UpdatedAt = r.GetInt64(r.GetOrdinal("updatedAt")),
            });
        }
        // A task left mid-run when the app was closed is not running any more.
        foreach (var t in list.Where(t => t.State is "running" or "checking"))
        {
            t.State = "idle";
            Upsert(t);
        }
        return list;
    }

    public void Upsert(LoopTask t)
    {
        t.UpdatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        using var cmd = Cmd("""
            INSERT INTO loop_task (id, title, prompt, cwd, doneWhen, checkKind, maxPasses,
                                   state, lastAttempt, updatedAt)
            VALUES (@p0,@p1,@p2,@p3,@p4,@p5,@p6,@p7,@p8,@p9)
            ON CONFLICT(id) DO UPDATE SET title=excluded.title, prompt=excluded.prompt,
              cwd=excluded.cwd, doneWhen=excluded.doneWhen, checkKind=excluded.checkKind,
              maxPasses=excluded.maxPasses, state=excluded.state,
              lastAttempt=excluded.lastAttempt, updatedAt=excluded.updatedAt
            """, t.Id, t.Title, t.Prompt, t.Cwd, t.DoneWhen,
                 t.Check == CheckKind.Shell ? "shell" : "agent",
                 t.MaxPasses, t.State, t.LastAttempt, t.UpdatedAt);
        cmd.ExecuteNonQuery();
    }

    public void Delete(string id)
    {
        using (var a = Cmd("DELETE FROM proof WHERE taskId = @p0", id)) a.ExecuteNonQuery();
        using (var b = Cmd("DELETE FROM loop_task WHERE id = @p0", id)) b.ExecuteNonQuery();
    }

    public void ClearProofs(string taskId)
    {
        using var cmd = Cmd("DELETE FROM proof WHERE taskId = @p0", taskId);
        cmd.ExecuteNonQuery();
    }

    public void Add(Proof p)
    {
        using var cmd = Cmd("""
            INSERT INTO proof (taskId, attempt, passed, makerOutput, checkerOutput, at)
            VALUES (@p0,@p1,@p2,@p3,@p4,@p5)
            """, p.TaskId, p.Attempt, p.Passed ? 1 : 0, p.MakerOutput, p.CheckerOutput, p.At);
        cmd.ExecuteNonQuery();
    }

    public List<Proof> Proofs(string taskId)
    {
        var list = new List<Proof>();
        using var cmd = Cmd("SELECT * FROM proof WHERE taskId = @p0 ORDER BY attempt", taskId);
        using var r = cmd.ExecuteReader();
        while (r.Read())
        {
            list.Add(new Proof(taskId, r.GetInt32(r.GetOrdinal("attempt")),
                r.GetInt32(r.GetOrdinal("passed")) == 1,
                r.GetString(r.GetOrdinal("makerOutput")),
                r.GetString(r.GetOrdinal("checkerOutput")),
                r.GetInt64(r.GetOrdinal("at"))));
        }
        return list;
    }
}
